docs/tutorial/graph_api.rst

```diff
@@ -0,0 +1,611 @@
+  :language: cuda
+
+Notice how each batch size equals our stream count — this ensures every stream stays busy.
+
+Synchronization
+^^^^^^^^^^^^^^^
+
+Each projection processes independently, so you only need to synchronize once at the end.
+:cpp:func:`hipStreamWaitEvent()` function makes the first stream wait for all other streams to complete.
+
+.. literalinclude:: ../tools/example_codes/graph_api_tutorial_main_streams.hip
+  :start-after: // [sphinx-sync-start]
+  :end-before: // [sphinx-sync-end]
+  :language: cuda
+
+Exploring the processing pipeline
+---------------------------------
+
+Now let us examine what happens to each projection. Find the ``START HERE`` comment to see the reconstruction pipeline's
+first steps:
+
+.. literalinclude:: ../tools/example_codes/graph_api_tutorial_main_streams.hip
+  :start-after: // [sphinx-preprocessing-start]
+  :end-before: // [sphinx-preprocessing-end]
+  :language: cuda
+
+This is a typical pattern found across many HIP applications: multiple kernels executing in sequence with data
+dependencies. In the next step, the weighted projections need to be transformed into Fourier space and filtered. For
+optimal performance, it is recommended to execute a 1D FFT on a buffer size which is a power of two. We therefore copy
+the weighted projection to another buffer where the row length is a power of two equal to or larger than the
+projection's row length:
+
+.. literalinclude:: ../tools/example_codes/graph_api_tutorial_main_streams.hip
+  :start-after: // [sphinx-proj-to-expanded-start]
+  :end-before: // [sphinx-proj-to-expanded-end]
+  :language: cuda
+
+Next, we transform the expanded projection into Fourier space for filtering:
+
+.. literalinclude:: ../tools/example_codes/graph_api_tutorial_main_streams.hip
+  :start-after: // [sphinx-forward-start]
+  :end-before: // [sphinx-forward-end]
+  :language: cuda
+
+.. tip::
+  Some hipFFT operations are graph-safe: As long as these operations are operating on the capturing stream, they will
+  be captured into the graph as well. Refer to :ref:`hipFFT's documentation <hipfft:hipfft-api-usage>` for more
+  information on its graph-safe operations.
+
+In Fourier space, we apply the Shepp-Logan filter, then transform back:
+
+.. literalinclude:: ../tools/example_codes/graph_api_tutorial_main_streams.hip
+  :start-after: // [sphinx-filter-start]
+  :end-before: // [sphinx-filter-end]
+  :language: cuda
+
+Shrink to original size and normalize the FFT output:
+
+.. literalinclude:: ../tools/example_codes/graph_api_tutorial_main_streams.hip
+  :start-after: // [sphinx-expanded-to-proj-start]
+  :end-before: // [sphinx-expanded-to-proj-end]
+  :language: cuda
+
+Finally, backproject the filtered projection into the 3D volume using ``atomicAdd`` operations to accumulate voxel
+values from multiple kernels:
+
+.. literalinclude:: ../tools/example_codes/graph_api_tutorial_main_streams.hip
+  :start-after: // [sphinx-bp-start]
+  :end-before: // [sphinx-bp-end]
+
```

Comment:
Please add a note that most of the preprocessing kernels are short kernels compared to backprojection. You can also mention that the kernel complexity compared to the others.